cupy/_math/sumprod.py

```diff
@@ -507,7 +507,65 @@ def gradient(f, *varargs, axis=None, edge_order=1):
         return outvals
 
 
-# TODO(okuta): Implement ediff1d
+def ediff1d(arr, to_end=None, to_begin=None):
+    """
+    Calculates the difference between consecutive elements of an array.
+
+    Args:
+        arr (cupy.ndarray): Input array.
+        to_end (cupy.ndarray, optional): Numbers to append at the end
+            of the returend differences.
+        to_begin (cupy.ndarray, optional): Numbers to prepend at the
+            beginning of the returned differences.
+
+    Returns:
+        cupy.ndarray: New array consisting differences among succeeding
+        elements.
+
+    .. seealso:: :func:`numpy.ediff1d`
+    """
+    # to flattened array.
+    arr = cupy.asanyarray(arr).ravel()
+
+    # to ensure the dtype of the output array is same as that of input.
+    dtype_req = arr.dtype
+
+    # if none optional cases are given
+    if to_begin is None and to_end is None:
+        return arr[1:] - arr[:-1]
+
+    if to_begin is None:
+        l_begin = 0
+    else:
+        to_begin = cupy.asanyarray(to_begin)
+        if not cupy.can_cast(to_begin, dtype_req, casting="same_kind"):
+            raise TypeError("dtype of `to_begin` must be compatible "
+                            "with input `arr` under the `same_kind` rule.")
+
+        to_begin = to_begin.ravel()
+        l_begin = len(to_begin)
+
+    if to_end is None:
+        l_end = 0
+    else:
+        to_end = cupy.asanyarray(to_end)
+        if not cupy.can_cast(to_end, dtype_req, casting="same_kind"):
+            raise TypeError("dtype of `to_end` must be compatible "
+                            "with input `arr` under the `same_kind` rule.")
+
+        to_end = to_end.ravel()
+        l_end = len(to_end)
+
+    # calulating using in place operation
+    l_diff = max(len(arr) - 1, 0)
+    result = cupy.empty(l_diff + l_begin + l_end, dtype=arr.dtype)
+    # result = arr.__array_wrap__(result)
```

Comment:
As CuPy does not support `__array_wrap__` protocol. Other way can be using `__array_ufunc__` protocol. Will it be wiser to add that here?

Reply:
We can skip calling `__array_wrap__` here as CuPy does not support subclassing `cupy.ndarray` yet, which may be considered in #3972. Would you leave a comment for now?

```suggestion
    # result = arr.__array_wrap__(result)  # CuPy does not support subclassing a ndarray
```

Reply:
Added!
Thanks for sharing the ongoing issue.